Paper: A New Python API for Webots Robotic Simulations

Justin-Fisher opened the pull request:
If you are creating this PR in order to submit a draft of your paper,
see http://procbuild.scipy.org/ for logs generated by the build
process.

See the [project readme](https://github.com/scipy-conference/scipy_proceedings#build-your-paper)
for more information.


*Editor*: <!--editor-->@stargaser<!--end-editor-->

*Reviewers*: <!--reviewers-list-->@matthewfeickert<!--end-reviewers-list-->

Justin-Fisher committed:
Add files via upload

Justin-Fisher committed:
Delete figure1.png

Justin-Fisher committed:
Delete justin_fisher.rst

Justin-Fisher committed:
Add files via upload

Justin-Fisher committed:
Delete justin_fisher.rst

Justin-Fisher committed:
Delete figure1.png

Justin-Fisher committed:
Add files via upload

Justin-Fisher committed:
Add files via upload

Justin-Fisher committed:
Merge branch 'scipy-conference:2022' into master

TelosTelos committed:
2022 submission

Justin-Fisher committed:
Delete figure1.png

from earlier year

Justin-Fisher committed:
Delete figures.pptx

from earlier year

Justin-Fisher committed:
Delete ~$figures.pptx

from earlier year

scoobies commented:
Hi!, I'm @scoobies, a friendly bot.

Type ```@scoobies help``` to discover how I can help you.

TelosTelos committed:
typo

TelosTelos committed:
Merge remote-tracking branch 'origin/master'

Justin-Fisher commented:
@scoobies build paper

scoobies commented:
building paper...

scoobies commented:
{"message"=>"Invalid paper specified", "status"=>"fail"}

Justin-Fisher commented:
@scoobies build status

scoobies commented:
{"Justin-Fisher-master"=>{"data"=>{"build_output"=>"Invalid paper"}, "status"=>"fail"}}

TelosTelos committed:
made title line length match title length

Justin-Fisher commented:
I'm at a loss here.  The paper looks fine in reST viewers I've tried.  Trying to build this using the `make_paper.sh` script produces an `output` folder but no `.pdf` and no error messages that I can see (though it does seem to flicker another terminal window very briefly on the screen as it runs, much too fast for me to read anything).  Trying to build it using `build_paper.py` produces an error saying `'pdflatex' is not recognized as an internal or external command` though I do have it installed, which seems like it'd be a path issue on my computer, though standard solutions to that haven't solved it yet.  The bot "scoobies" here seems to think the paper is invalid, but won't say why.  It hasn't appeared on the build server page, so I can't see what errors that might think it has.

One hypothesis might be that I think I used this same master branch to submit something to SciPy many years ago (circa 2016). Not sure if that's confusing it somehow?

mepa commented:
@scoobies build paper

scoobies commented:
building paper...

scoobies commented:
{"data"=>{"info"=>"Build for paper Justin-Fisher-master scheduled.  Note that builds are only executed if the current build attempt is more than 5 minutes old."}, "status"=>"success"}

mepa commented:
Hi @Justin-Fisher, the PR needs to have a title that begins with "Paper:" in order for the build system to find it. Sorry the errors aren't good at pointing to that! The paper built successfully after I updated the PR title.

deniederhut commented:
@scoobies mark paper

deniederhut commented:
@scoobies build status

scoobies commented:
{"Justin-Fisher-master"=>{"data"=>{"build_output"=>"[*] Updating proceedings build tools in {self.master_repo_path}...\nAlready up to date.\n[*] Check out paper repository...\nCloning into '/tmp/tmpoosf_hlu'...\nMoving proceedings build tools to temp directory...\n[*] Build the paper...\nBuilding: justin_fisher\n*** Warning: /tmp/tmpoosf_hlu/publisher/../output/justin_fisher/paper_stats.json does not exist.\n", "build_pdf_path"=>"/procbuild/procbuild/../cache/Justin-Fisher-master.pdf", "build_status"=>"success", "build_timestamp"=>"29/05 01:39"}, "status"=>"success"}}

deniederhut commented:
@scoobies mark needs review

deniederhut commented:
@scoobies assign @stargaser as editor

scoobies commented:
Assigned! @stargaser is now the editor

TelosTelos committed:
added code readability metrics, typo fixes

matthewfeickert commented:
:wave: Hi @Justin-Fisher, I'll be one of the reviewers for your paper. Expect an initial review in the form of PR comments by EOD 2022-06-09 at the latest. I'll be reviewing the current build up on https://procbuild.scipy.org/ from 2022-05-30, so please ping me here as reminder if you make any revisions in the meantime.

mepa commented:
@scoobies add @matthewfeickert to reviewers

scoobies commented:
@matthewfeickert added to the reviewers list!

deniederhut commented:
@scoobies add @zuhalcakir to reviewers

scoobies commented:
@zuhalcakir added to the reviewers list!

deniederhut commented:
Hi! We've moved the deadline for first reviews back one week, to June 16th, to give us a little more time to find reviewers for all the papers this year. You can see the updated full timeline at https://github.com/scipy-conference/scipy_proceedings#timeline-for-2022.

matthewfeickert commented:
@Justin-Fisher, I've done a first round of review, but the May 27th 1st draft submission was supposed to be finished drafts that could be reviewed in full. Your paper is currently incomplete and lacks any conclusion or references.

To avoid creating additional work of multiple reviews can you please give @stargaser, @zuhalcakir, and myself a timeline on when you will provide a finished draft that can be properly reviewed?

deniederhut commented:
It looks like some of the tables aren't rendering properly too. @stargaser have you had a chance to look at this?

TelosTelos committed:
readability refs, short conclusion, table reformat

Justin-Fisher commented:
Okay, I've added the references mentioned in the old draft.

I've added a short conclusions section, though I'm open to guidance regarding what sort of conclusion is appropriate for a paper like this?

And I've fiddled with table settings to mostly solve the format issues it was having with tables spilling across columns, though I do still see some issues where it line-wraps too much or too little -- will keep fiddling.

deniederhut commented:
Hey @Justin-Fisher ! There is wide table class directive you can use for tables that need to break column boundaries. There is an example of that here: https://raw.githubusercontent.com/scipy-conference/scipy_proceedings/2022/papers/00_bibderwalt/00_bibderwalt.rst

matthewfeickert commented:
> Okay, I've added the references mentioned in the old draft.
> 
> I've added a short conclusions section, though I'm open to guidance regarding what sort of conclusion is appropriate for a paper like this?

I'll review this before the 16th and give feedback as a PR review along with the notes I have so far. The proceedings should be treated as short scientific papers, so conclusions sections are expected but they can be quite short. The [SciPy proceedings are public](https://conference.scipy.org/proceedings/) so if you're unsure how you want to write your conclusions and need some inspiration you have plenty of published proceedings as examples.

deniederhut commented:
Hey @matthewfeickert and @zuhalcakir! How are the reviews coming along?

matthewfeickert commented on justin_fisher.rst:
```diff
@@ -0,0 +1,242 @@
+:author: Justin C. Fisher
```
Is there a reason this paper is here? It is not the paper that is being reviewed and should be removed.

stargaser replied:
@Justin-Fisher any files outside the papers subdirectory must be removed from the pull request, otherwise they interfere with the build of the proceedings.

matthewfeickert replied:
This file still needs to be removed.

matthewfeickert commented on papers/justin_fisher/justin_fisher.rst:
```diff
@@ -0,0 +1,217 @@
+1. Introduction
+---------------
+
+Webots is a popular open-source package for 3D robotics simulations. [Mic01]_ It can also be used as a 3D interactive environment for other physics-based modeling, virtual reality, teaching or games. Webots has historically provided a simple Python API, allowing Python programs to control individual robots or the simulated world. This Python API was a thin wrapper over a C++ API, which itself was ...
```
(Having to make paragraph level comments given all 1 line. :( )

> Webots is a popular open-source package for 3D robotics simulations

In addition to the IJARS citation here it would be nice to get a code citation for Webots as well.

> Furthermore, this API was not very "pythonic" and did not provide many of the conveniences that help to make development in Python be fast, intuitive, and easy to learn.

This is a bit strange to write in the past tense. The C-like API is still the existing API. Change tense to present.

matthewfeickert commented on papers/justin_fisher/justin_fisher.rst:
```diff
@@ -0,0 +1,217 @@
+
+Webots is a popular open-source package for 3D robotics simulations. [Mic01]_ It can also be used as a 3D interactive environment for other physics-based modeling, virtual reality, teaching or games. Webots has historically provided a simple Python API, allowing Python programs to control individual robots or the simulated world. This Python API was a thin wrapper over a C++ API, which itself was ...
+
+In qualitative terms, the old API felt like one is awkwardly using Python to call C and C++ functions, whereas the new API feels much simpler, much easier, and like it was fully intended for Python.  Here is a representative (but far from comprehensive) list of examples:
```
This sentence doesn't really add much. Can just make it clear that the current Python API is a direct port of a `C`-like API and that the new API is more Pythonic. Be careful with unqualified use of easier &mdash; easier for what? It is clear (from the [C++/Java/Python user guide examples](https://cyberbotics.com/doc/guide/cpp-java-python#python-example)) that Webots is providing a stripped down SDK and that they have decided to not provide "localization" of the SDK languages via idiomatic APIs and so the APIs are all `C`-like. You'll show us in the rest of the paper why the new API is more Pythonic, so maybe just focus on that message.

matthewfeickert commented on papers/justin_fisher/justin_fisher.rst:
```diff
@@ -0,0 +1,217 @@
+
+* Unlike the old API, the new API contains helpful Python type annotations and docstrings.
+* Webots employs many vectors, e.g., for 3D positions, 4D rotations, and RGB colors.  The old API typically treated these as lists or integers (24-bit colors).  In the new API these are Vector objects, with conveniently addressable components (e.g. :code:`vector.x` or :code:`color.red`), convenient helper methods like :code:`vector.magnitude` and :code:`vector.unit_vector`, and overloaded vector  ...
+* The new API also provides easy interfacing between high-resolution Webots sensors (like cameras and Lidar) and Numpy arrays, to make it much more convenient to use Webots with popular python packages like Numpy, Scipy, PIL or OpenCV.  For example, converting a Webots camera image to a NumPy array is now as simple as :code:`camera.array` and this now allows the array to share memory with the cam ...
```
> popular python packages like Numpy, Scipy, PIL or OpenCV

These should all be cited the first time they appear in the paper and "python" should be capitalized as it is the language not the interpreter.

matthewfeickert commented on papers/justin_fisher/justin_fisher.rst:
```diff
@@ -0,0 +1,217 @@
+* Most attributes are now accessible (and alterable, when applicable) by pythonic properties like :code:`motor.velocity`.
+* Many devices now have python methods like :code:`__bool__` overloaded in intuitive ways.  E.g., you can now use :code:`if bumper` to detect if a bumper has been pressed, rather than the old :code:`if bumper.getValue()`.
+* Pythonic container-like interfaces are now provided.  You may now use :code:`for target in radar` to iterate through the various targets a radar device has detected or :code:`for packet in receiver` to iterate through communication packets that a receiver device has received (and it now automatically handles a wide variety of python objects, not just strings).
+* The old API required supervisor controllers to use a wide variety of separate functions to traverse and interact with the simulation’s scene tree, including different functions for different VRML datatypes (like :code:`SFVec3f` or :code:`MFInt32`). The new API automatically invisibly handles these datatypes and translates intuitive python syntax (like dot-notation and square-bracket indexing) t ...
```
> The new API automatically invisibly handles these datatypes

what do you mean by "invisibly"? Do you mean automatically?

> translates intuitive python syntax

"python" should be capitalized as "Python" as it is the language not the interpreter.

matthewfeickert commented on papers/justin_fisher/justin_fisher.rst:
```diff
@@ -0,0 +1,217 @@
+* Pythonic container-like interfaces are now provided.  You may now use :code:`for target in radar` to iterate through the various targets a radar device has detected or :code:`for packet in receiver` to iterate through communication packets that a receiver device has received (and it now automatically handles a wide variety of python objects, not just strings).
+* The old API required supervisor controllers to use a wide variety of separate functions to traverse and interact with the simulation’s scene tree, including different functions for different VRML datatypes (like :code:`SFVec3f` or :code:`MFInt32`). The new API automatically invisibly handles these datatypes and translates intuitive python syntax (like dot-notation and square-bracket indexing) t ...
+
+As another example illustrating how much easier the new API is to use, here are two lines from Webots' sample :code:`supervisor_draw_trail`, as it would appear in the old Python controller.
```
> here are two lines from Webots' sample :code:`supervisor_draw_trail`

This should get a citation or should link to a versioned doc example of this.

matthewfeickert commented on papers/justin_fisher/justin_fisher.rst:
```diff
@@ -0,0 +1,217 @@
+.. code-block:: python
+
+  root_children_field = supervisor.getField(supervisor.getRoot(), "children")
+  root_children_field.importMFNodeFromString(-1, trail_plan)
```
This code block needs to be reformatted to fit on the page properly in the paper build.

matthewfeickert commented on papers/justin_fisher/justin_fisher.rst:
```diff
@@ -0,0 +1,217 @@
+
+The new API is mostly backwards-compatible with the old Python Webots API, and provides an option to display deprecation warnings with helpful advice for changing to the new API.
+
+The new Python API is planned for inclusion in an upcoming Webots release, to replace the old one.  In the meantime, an early-access version is available, distributed with the same liberal Apache open-source license as Webots itself.
```
> The new Python API is planned for inclusion in an upcoming Webots release

Can you say which one it is planned for (if this is currently known)?

> distributed with the same liberal Apache open-source license as Webots itself.

state explicitly that this is "Apache License 2.0". So
> distributed under Apache License 2.0, the same license that Webots is distributed under.

matthewfeickert commented on papers/justin_fisher/justin_fisher.rst:
```diff
@@ -0,0 +1,217 @@
+
+The new Python API is planned for inclusion in an upcoming Webots release, to replace the old one.  In the meantime, an early-access version is available, distributed with the same liberal Apache open-source license as Webots itself.
+
+In what follows, the history and motivation for this new API is discussed, including its use in teaching an interdisciplinary undergraduate Cognitive Science course called Minds, Brains and Robotics.  Some of the design decisions for the new API are discussed, which will not only aid in understanding it, but also have broader relevance to parallel dilemmas that face many other software developers ...
```
> interdisciplinary undergraduate Cognitive Science course called Minds, Brains and Robotics

Please add a citation (also cool course! :rocket:)

> And some metrics are given to quantify how the new API has improved over the old.

This is a bit vague. Maybe drop the "And some" and clarify what areas you are considering "improved".

matthewfeickert commented on papers/justin_fisher/justin_fisher.rst:
```diff
@@ -0,0 +1,217 @@
+
+This interdisciplinary class draws students with diverse backgrounds, and programming skills. Accomodating those with fewer skills required simplifying many of the complexities of the old Webots API.  It also required setting up tools to use Webots "supervisor" powers to help manipulate the simulated world, e.g. to provide students easier customization options for their robots.  The old Webots AP ...
+
+[Not sure whether to include more examples or images here of ways in which the new API were relevant to this class?]
```
Remove this from the paper. But yes, is probably the answer.

matthewfeickert commented on papers/justin_fisher/justin_fisher.rst:
```diff
@@ -0,0 +1,217 @@
+
+Reduction of wordiness and punctuation helps to make programs easier to read and to understand, and it reduces the cognitive load on coders.  However, there are also drawbacks.
+
+One drawback is that, properties can give the mistaken impression that some attributes are computationally cheap to get or set. In cases where this impression would be misleading, more traditional method calls were retained and/or the comparative expense of the operation was clearly documented.
```
```suggestion
One drawback is that properties can give the mistaken impression that some attributes are computationally cheap to get or set. In cases where this impression would be misleading, more traditional method calls were retained and/or the comparative expense of the operation was clearly documented.
```

erroneous comma

matthewfeickert commented on papers/justin_fisher/justin_fisher.rst:
```diff
@@ -0,0 +1,217 @@
+
+Two other drawbacks are related.  One is that inviting ordinary users to assign properties to API objects might lead them to assign other attributes that could cause problems. Since Python lacks true privacy protections, it has always faced this sort of worry, but this worry becomes even worse when users start to feel familiar moving beyond just using defined methods to interact with an object.
+
+Relatedly, Python debugging provides direct feedback in cases where a user misspells :code:`motor.setFoo(v)` but not when someone mispells 'motor.foo = v`.  If a user inadvertently types :code:`motor.setFool(v)` they will get an :code:`AttributeError` noting that :code:`motor` lacks a :code:`setFool` attribute.  But if a user inadvertently types :code:`motor.fool = v`, then Python will silently c ...
```
> But if a user inadvertently types :code:`motor.fool = v`, then Python will silently create a new :code:`.fool` attribute for :code:`motor` and the user will often have no idea what has gone wrong.

Why not use `__slots__` ? https://docs.python.org/3/reference/datamodel.html?highlight=slots#slots

```python
>>> class A():
...     __slots__ = ["foo"]
...
>>> motor = A()
>>> dir(motor)
['__class__', '__delattr__', '__dir__', '__doc__', '__eq__', '__format__', '__ge__', '__getattribute__', '__gt__', '__hash__', '__init__', '__init_subclass__', '__le__', '__lt__', '__module__', '__ne__', '__new__', '__reduce__', '__reduce_ex__', '__repr__', '__setattr__', '__sizeof__', '__slots__', '__str__', '__subclasshook__', 'foo']
>>> motor.foo = 8
>>> motor.foo
8
>>> motor.fool = 8
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'A' object has no attribute 'fool'
```

matthewfeickert commented on papers/justin_fisher/justin_fisher.rst:
```diff
@@ -0,0 +1,217 @@
+
+3.2 Backwards Compatibility.
+============================
+The new API offers many new ways of doing things, ways that are often better by most objective metrics, with the main drawback being just that they differ from old ways.  The possibility of making a clean break from the old API was considered, but that would stop old code from working, alienate veteran users, and risk causing a schism akin to the deep one between Python 2 and Python 3 communities ...
```
> The new API offers many new ways of doing things, ways that are often better by most objective metrics, with the main drawback being just that they differ from old ways

You have yet to actually show any objective metrics at this point and this language is pretty broad for the scope of the paper. I would suggest narrowing what you mean by "better" or getting more detailed in your description.

> and risk causing a schism akin to the deep one between Python 2 and Python 3 communities.

I'm not sure it is very reasonable to say that there are modern "Python 2 and Python 3 communities." Python 2 has been EOL since 2020-01-01. There will be plenty of Python 2 users for a long time because of legacy code (experiments that I work on among them) but I think one should be careful using this verbage 2+ years after the EOL. There _was_ a deep schism between Python 2 users and Python 3 users in the past until around Python 3.6 or so, but if you have clear examples in mind it is probably best to state them specifically. I might just be misunderstanding what you mean here, so clarification is great too.

matthewfeickert commented on papers/justin_fisher/justin_fisher.rst:
```diff
@@ -0,0 +1,217 @@
+
+Another option would have been to refrain from adding a new-and-better feature to avoid introducing redundancies or backward incompatibilities. But that has obvious drawbacks too.
+
+Instead, a compromise was typically adopted: to provide both the new-and-better way and the worse-old way.  This redundancy was eased by shifting from :code:`getFoo` / :code:`setFoo` methods to properties, and from :code:`CamelCase` to Pythonic :code:`snake_case`, which reduced the number of name collisions between old and new.   Employing the worse old way leads to a deprecation warning that inc ...
```
> Instead, a compromise was typically adopted: to provide both the new-and-better way and the worse-old way

Again, better and worse are used here in subjective ways about API design. Consider more descriptive and less editorial terms like `C`-like and Pythonic.

matthewfeickert commented on papers/justin_fisher/justin_fisher.rst:
```diff
@@ -0,0 +1,217 @@
+
+Brief inspection should reveal that the latter code is much easier to read, write and understand, not just because it is shorter, but also because its punctuation is limited to standard Python syntax for traversing attributes of objects, because it reduces the need to introduce new variables like :code:`trail_node` for things that it already makes easy to reference (via :code:`world.TRAIL`, which ...
+
+This intuitive impression is confirmed by automated metrics for code readability.  The measures below consider the full :code:`supervisor_draw_trail` sample controller (from which the above snippet was drawn), since this is the Webots sample controller that makes the most sustained use of supervisor functionality to perform a fairly plausible supervisor task (maintaining the position of a streame ...
```
> The measures below consider the full :code:`supervisor_draw_trail` sample controller

In general "below" isn't very helpful in typeset papers, especially when you can directly link to the tables that you mean and alleviate any ambiguity.

> (Sample code and computations of metrics are available under additional information below.)

These should all be citations instead (that preserve the links to specific tags).

matthewfeickert commented on papers/justin_fisher/justin_fisher.rst:
```diff
@@ -0,0 +1,217 @@
+  |CC Cyclomatic Complexity                               | 5 (Grade A) | 8 (Grade B)  |
+  +-------------------------------------------------------+-------------+--------------+
+
+Some raw measures for the two controllers are shown in Table :ref:`metrictable`. These were gathered using the Radon code-analysis tools. The "lines of code" measures reflect that the new API makes it easier to do more things with less code. (The measures differ in how they count blank lines, comments, multi-line statements, and multi-statement lines like :code:`if p: q()`.)  Line counts can be m ...
```
> These were gathered using the Radon code-analysis tools

Cite this.

matthewfeickert commented on papers/justin_fisher/justin_fisher.rst:
```diff
@@ -0,0 +1,217 @@
+
+This intuitive impression is confirmed by automated metrics for code readability.  The measures below consider the full :code:`supervisor_draw_trail` sample controller (from which the above snippet was drawn), since this is the Webots sample controller that makes the most sustained use of supervisor functionality to perform a fairly plausible supervisor task (maintaining the position of a streame ...
+
+.. table:: Length and Complexity Metrics. :label:`metrictable`
```
All tables and figures need proper captions. Also why

> LoC Lines of Code

instead of 

> Lines of Code (LoC)

? The latter is clearer.

matthewfeickert replied:
The tables are formatted much nicer now (I fully appreciate that this can be a bit of a pain when fighting with a typesetting engine, so thanks!) though it would be great if there could be some additional context in the captions of the tables. At the moment the captions are just acting as table titles, which isn't necessary.

matthewfeickert commented on papers/justin_fisher/justin_fisher.rst:
```diff
@@ -0,0 +1,217 @@
+
+Another collection of classic measures of code readability was developed by Halstead. [Hal01]_ These measures (especially volume) have been shown to correlate with human assessments of code readability [Bus01]_ [Pos01]_ These measures generally penalize a program for using a "vocabulary" involving more operators and  operands. Table :ref:`halsteadtable` shows these metrics, as computed by Radon.  ...
+
+.. table:: Halstead Metrics. :label:`halsteadtable`
```
All tables need proper captions and to be formatted properly so they fit on a page.

matthewfeickert commented on papers/justin_fisher/justin_fisher.rst:
```diff
@@ -0,0 +1,217 @@
+
+Lastly, the Maintainability Index, and variants thereof, are a measure of how easy to support and change source code is. [Oman01]_ Variants of the Maintainability Index are commonly used, including in Microsoft Visual Studio. These measures combine Halstead Volume, Source Lines of Code, and Cyclomatic Complexity, all mentioned above, and two variants (SEI and Radon) also provide credit for percen ...
+
+.. table:: Maintainability Index Metrics. :label:`maintaintable`
```
All tables need proper captions and to be formatted properly so they fit on a page.

matthewfeickert commented on papers/justin_fisher/justin_fisher.rst:
```diff
@@ -0,0 +1,217 @@
+
+The Cyclomatic Complexity score counts the number of potential branching points that appear within the code, like :code:`if`, :code:`while` and :code:`for`. [McC01]_ Cyclomatic Complexity is strongly correlated with other plausible measures of code readability involving indentation structure. [Hin01]_ The new API's score is lower/better due to its automatically converting vector-like values to th ...
+
+Another collection of classic measures of code readability was developed by Halstead. [Hal01]_ These measures (especially volume) have been shown to correlate with human assessments of code readability [Bus01]_ [Pos01]_ These measures generally penalize a program for using a "vocabulary" involving more operators and  operands. Table :ref:`halsteadtable` shows these metrics, as computed by Radon.  ...
```
```suggestion
Another collection of classic measures of code readability was developed by Halstead. [Hal01]_ These measures (especially volume) have been shown to correlate with human assessments of code readability [Bus01]_ [Pos01]_. These measures generally penalize a program for using a "vocabulary" involving more operators and  operands. Table :ref:`halsteadtable` shows these metrics, as computed by Radon. The new API scores significantly lower/better on these metrics, due in large part to its invisibly selecting among many different C-API calls without these needing to appear in the user's code.  E.g. having :code:`motor.velocity` as a unified property involves fewer unique names than having users write both :code:`setVelocity()` and :code:`getVelocity()`, and often forming a third local :code:`velocity` variable.  And having :code:`world.children[-1]` access the last child that field in the simulation saves having to count :code:`getField`, and :code:`getMFNode` in the vocabulary, and often also saves forming additional local variables for nodes or fields gotten in this way.  Both of these factors also help the new API to greatly reduce parentheses counts.
```

Missing a period.

matthewfeickert commented on papers/justin_fisher/justin_fisher.rst:
```diff
@@ -0,0 +1,217 @@
+
+.. [McC01] McCabe, TJ. "A Complexity Measure" , 2(4): 308-320. 1976.
+
+.. [Mic01] Michel, O. "Webots: Professional Mobile Robot Simulation. *Journal of Advanced Robotics Systems.* 1(1): 39-42. 2004.  http://www.ars-journal.com/International-Journal-of-Advanced-Robotic-Systems/Volume-1/39-42.pdf
```
> http://www.ars-journal.com/International-Journal-of-Advanced-Robotic-Systems/Volume-1/39-42.pdf

This url does not point to a paper. It resolves to a page with plain text that reads "not file".

matthewfeickert commented on papers/justin_fisher/justin_fisher.rst:
```diff
@@ -0,0 +1,217 @@
+
+Radon tool used to compute code readability metrics: https://radon.readthedocs.io/en/latest/index.html
+
+References
+==========
```
Do none of your references have DOIs? If they do have DOIs please include them.

matthewfeickert commented on papers/justin_fisher/justin_fisher.rst:
```diff
@@ -0,0 +1,217 @@
+
+There are potential concerns about each of these measures of code readability, and one can easily imagine playing a form of "code golf" to optimize some of these scores without actually improving readability (though it would be difficult to do this for all scores at once). Fortunately, most plausible measures of readability have been observed to be strongly correllated across ordinary cases, [Pos ...
+
+[Could include computational performance metrics as well?  Probably the best tests would be (a) transmission of high-bandwidth devices like Camera images, and (b) transmission of numerous supervisor control signals.]
```
```suggestion
```

Remove, but yes.

matthewfeickert commented on papers/justin_fisher/justin_fisher.rst:
```diff
@@ -0,0 +1,217 @@
+
+A new Python API for Webots robotic simulations was presented. It more efficiently interfaces directly with the Webots C API and provides a more intuitive, easily usable, and "pythonic" interface for controlling Webots robots and simulations. Motivations for the API and some of its design decisions were discussed.  Advantages of the new API were discussed and quantified using automated code reada ...
+
+[Not sure this section was needed?]
```
```suggestion
```

Remove, but yes all papers need conclusions.

matthewfeickert commented on papers/justin_fisher/justin_fisher.rst:
```diff
@@ -0,0 +1,217 @@
+
+[Could include computational performance metrics as well?  Probably the best tests would be (a) transmission of high-bandwidth devices like Camera images, and (b) transmission of numerous supervisor control signals.]
+
+5. Conclusions
+==============
+
+A new Python API for Webots robotic simulations was presented. It more efficiently interfaces directly with the Webots C API and provides a more intuitive, easily usable, and "pythonic" interface for controlling Webots robots and simulations. Motivations for the API and some of its design decisions were discussed.  Advantages of the new API were discussed and quantified using automated code reada ...
```
This is very light on detail for a conclusions section.

matthewfeickert reviewed (changes requested):
Thanks for a very interesting proceedings submission @Justin-Fisher! I think that you have a paper that is mostly there and with some short rounds of revision will be a very nice contribution to the proceedings.

There are unfortunately quite a bit of formatting errors that will need revision (an excellent example of a well formatted paper that deals with complex figures and image placing is [signac: Data Management and Workflows for Computational Researchers](https://conference.scipy.org/proceedings/scipy2021/bradley_dice.html) from SciPy 2021) and as a reminder submissions are supposed to be finished proceedings that if they pass review could be accepted (i.e. notes to yourself in the paper should be removed).

This review follows the [Suggested Review Criteria for SciPy Proceedings](https://github.com/scipy-conference/scipy_proceedings/blob/632099a7e62d192a5d66156dc70547fd9f35cb81/review_criteria.md).

## Written Quality

- [x] Prose should be written in English.
- [ ] Prose should clearly and consistently communicate the narrative (e.g. consistent tense and person; use of active voice; etc)
   * There is inconsistent casing of "Pythonic" (switches between lowercase and uppercase).
   * There is some flipping between tenses (e.g. "is" vs. "was") so just pick a tense and stick with it.
- [x] The written style should convey information that furthers the knowledge or research of the reader.
- [ ] Due to the interdisciplinary nature of SciPy, highly domain-specific jargon should be avoided or explained where possible.
   * It would help to explain more what the "simulated world" entails and how much detail exists. This would provide more context for the code examples.

## Technical Content

- [ ] The technical content should be scientifically sound.
   * This is mostly there, but there is quite a bit of subjective language (e.g. "good" vs. "bad") used throughout that is expressing opinion and not performance. More importantly, when choosing code metrics it is important to define _what_ you are trying to measure with them and _why_. It is clear that the metric that is being optimized for is readability, but the metric comparison tables lack sufficient captions and scope to identify if the reported differences are substantial. Table 1 is not convincing and Table 2 does not identify which of the metrics are of interest, given that most of the metrics in the table are computed from Volume and Length (also `h1` and `h2` are not defined.) The focus on readability is fine, but it should be stated for comparison why performance metrics are not of interest for the paper or real world performance metrics could be added. The author seems aware of this though in general given comments in the text, and so revisions or additions are probably not an undue burden for them. :+1:
- [x] Computational content should, likewise, be verified.
   - There is a link (should be a reference to a tagged version) to https://github.com/Justin-Fisher/new_python_api_for_webots that does provide a [`measure_code_complexity.py` for the `supervisor_draw_rail_python` example](https://github.com/Justin-Fisher/new_python_api_for_webots/blob/d180bcc7f505f8168246bee379f8067dfaf373ea/webots_new_python_api_samples/controllers/supervisor_draw_trail_python/measure_code_complexity.py). I have not yet stepped through the process of verifying the results as the installation is lightly documented (though for something experimental like this that's not meant to be real criticism) and the testing procedure is undocumented. The being said, there are comments that have been included in the code complexity script that indicate its results at last runtime, so while not ideal it is probably sufficient for the conference.
- [x] The work should describe the development or use of python software for approaching a problem in a domain within the scope of the conference.
   * Yes, it seems to be a nice use case combining robotics and education.

## Novelty and Impact

- [x] Assume that the novelty of the conceptual work described in the paper has been assessed by the Program Committee.
  The paper should advance the state of a scientific domain, the practice of scientific computing itself, or another subject area within the scope of the conference.
   * The paper provides work on a Pythonic API to a robotics simulator that provides an alternative to the C-like Python API/bindings that are currently provided. This work is substantial and combined with the author's use of it in University level educational settings it is well within the scope of the conference.

## Verifiability

- [ ] Software descriptions should be accompanied by references to or examples of representative source code.
   * The paper focuses on the `supervisor_draw_trail` example, but doesn't provide a reference to the source code for comparison as only a small example section of the Python API code is given.
- [x] Source code essential to the conclusions of the paper should be made available to the reader.
   * As I have yet to fully replicate the results I can't say with certainty that the source code is sufficient, but for the time being I am willing to say that as it exists publicly that it is hypothetically replicable.
- [ ] Data sources should be identified (e.g., with citation to a persistent DOI).
   * In this situation data include source code and dependencies. The provided source code should cited at a specific tag, and not just linked, and ideally be given a DOI (i.e. [with Zenodo](https://zenodo.org/)).
- [ ] Analysis should be accompanied by a workflow description sufficient to reproduce the results.
  * Analysis workflow instructions are not provided.

## Other Requirements

- [ ] All mentioned software should be formally and consistently cited wherever possible.
   * None of the software mentioned in the paper is properly cited.
- [x] Acronyms should be spelled out upon first mention.
- [x] License conditions on images and figures must be respected (Creative Commons, etc.).
- [x] Mathematical and other symbols should be defined.
- [x] Definitions should include consistent units where appropriate.
- [ ] Avoid long paragraphs of your rich text file to enable reviewers to target specific sections with their comments.
  GitHub comments are per line of text.
   * It would be nice if the paper followed something more like "1 sentence per line" to make code review style review like SciPy does more feasible.
- [x] Avoid custom LaTeX markup where possible.

### Length

- [x] The compiled version should be no longer than 8 pages, including figures, source code, appendices. References are excluded from this limit.

### Figures

- [ ] All figures and tables should have captions.
   * None of the tables have sufficient captions. Captions should provide sufficient information that a reader is able to understand the core components and relevance of the figure or table.
   * Additionally, there is table collision that needs to be resolved.
- [x] Figure text should be of a readable size.
- [x] Images and figures should be reasonably sized and formatted for viewing online; typically a few hundred kilobytes and less than 1 MB.

### Code Snippets

- [ ] Code snippets should be formatted to fit inside a single column without overflow. Please make sure to check the build server's copy of your paper; it's the authoritative view of how your paper will appear.
   * There are multiple formatting issues with the code snippets that need to be resolved. Some of them are bad enough that the code is unreadable.
- [x] If not in Python, the language in the code snippet should be mentioned.
- [x] Code snippets should follow a common style guide. PEP8 is preferred.
   * All the code snippets are incomplete segments and not actually runnable, so while it is unclear if a formatting style is being used it isn't very relevant for 1-liners.

stargaser commented:
@scoobies mark needs more review

deniederhut commented:
@scoobies remove @zuhalcakir from reviewers

scoobies commented:
@zuhalcakir removed from the reviewers list!

deniederhut commented:
@scoobies mark pending comment

deniederhut commented:
Hi @Justin-Fisher ! Have you had a chance to respond to Matthew's comments?

Justin-Fisher commented:
Thanks for the very helpful feedback @matthewfeickert  Sorry to have been struggling with some family health issues, which forced this briefly to a back burner.  I anticipate having changes finished within the next 24 hours.  For now, I'll provide some quick links in response to the following comment.

> The paper focuses on the supervisor_draw_trail example, but doesn't provide a reference to the source code for comparison as only a small example section of the Python API code is given.

You note that you found and read through the [`measure_code_complexity.py`](https://github.com/Justin-Fisher/new_python_api_for_webots/blob/d180bcc7f505f8168246bee379f8067dfaf373ea/webots_new_python_api_samples/controllers/supervisor_draw_trail_python/measure_code_complexity.py) at the first site listed under additional information.  The source code whose complexity was compared is in the same folder in the two files that are opened at the top of that code, namely  [`supervisor_draw_trail_old_api_bare_bones.py`](https://github.com/Justin-Fisher/new_python_api_for_webots/blob/d180bcc7f505f8168246bee379f8067dfaf373ea/webots_new_python_api_samples/controllers/supervisor_draw_trail_python/supervisor_draw_trail_old_api_bare_bones.py)  and  [`supervisor_draw_trail_new_api_bare_bones.py`](https://github.com/Justin-Fisher/new_python_api_for_webots/blob/d180bcc7f505f8168246bee379f8067dfaf373ea/webots_new_python_api_samples/controllers/supervisor_draw_trail_python/supervisor_draw_trail_new_api_bare_bones.py).  (Both have "bare bones" in their names because I eliminated some irrelevant clutter for the comparison like licensing boilerplate, and wrapped each in a function definition as is needed for Radon to measure code complexity.)

TelosTelos committed:
"pythonic" is now consistently lowercase, and present tense is now used for both old and new api

matthewfeickert commented:
> Thanks for the very helpful feedback @matthewfeickert Sorry to have been struggling with some family health issues, which forced this briefly to a back burner.

No worries &mdash; those are obviously the right priorities! Hope everything is okay.

TelosTelos committed:
added description of the potential complexity of Webots simulations to beginnings of sections 1 and 2

TelosTelos committed:
justified the presentation of multiple code-readability metrics without committing to which is "best". Fixed h1/h2 -> n1/n2

TelosTelos committed:
added carriage returns by sentence to make it easier to refer to edits by line on github.

TelosTelos committed:
Consistently capitalized Python when referring to the language.

Justin-Fisher commented:
Thanks for the good wishes, and thanks again for all the thoughtful suggestions!  I'll make a laundry-list of changes as I make them.  Further down is a remaining TO DO list (largely for my own benefit) and a short list of suggested changes that I'm not sure I'll make.

**Changes Made**
* Changed "pythonic" to be consistently lower-case
* Consistently capitalized Python when referring to the language.
* Changed discussion of old and new API to both use present tense.
* Added description of the potential amount of detail in Webots simulations to first section and background section.  Pasted here for convenience:
> Webots uses the Open Dynamics Engine (ODE), which allows physical simulations of Newtonian bodies, collisions, joints, springs, friction, and fluid dynamics.  Webots provides the means to simulate a wide variety of robot components, including motors, actuators, wheels, treads, grippers, light sensors, ultrasound sensors, pressure sensors, range finders, radar, lidar, and cameras (with many of these sensors drawing their inputs from GPU processing of the simulation).  A typical simulation will involve one or more robots, each with somewhere between 3 and 30 moving parts (though more would be possible), each running its own controller program to process information taken in by its sensors to determine what control signals to send to its devices.  A simulated world typically involves a ground surface (which may be a sloping polygon mesh) and dozens of walls, obstacles, and/or other objects, which may be stationary or moving in the physics simulation.

> For example, an early activity in this class involves building Braitenburg-style vehicles [Bra01]_ that use light sensors and cameras to detect a lamp carried by a hovering drone, as well as ultrasound and touch sensors to detect obstables.  Using these sensors, the robots navigate towards the lamp in a cluttered playground sandbox that includes sloping sand, an exterior wall, and various obstacles including a puddle of water and platforms from which robots may fall.
* Added a note explaining why multiple metrics are given without committing myself in debates regarding which is most relevant.  Pasted here for convenience
> Multiple metrics are reported because theorists disagree about which are most relevant in assessing code readability, because some of these play a role in computing other metrics discussed below, and because this may help to allay potential worries that a few favorable metrics might have been cherry-picked. This paper provides some explanation of these metrics and of their potential significance, while remaining neutral regarding which, if any, of these metrics is best.
* fixed references to h1, h2. They should have been n1, n2, defined before.  (They're often represented with Greek etas, and I wavered regarding which English letter to use.)
* added scare-quotes around "better" when indicating whether higher or lower scores on different readability metrics are conventionally taken to be "better"
* similarly added scare-quotes around "new-and-better" in the section on backwards compatibility and softened the claim that our new ways actually are "better". The main point of that section is just to discuss different approaches to backwards compatibility issues, issues that typically arise only when you have a new way of doing things that you think is "better" enough to want to introduce.
* Interesting and clever suggestion to use `__slots__` to help people catch cases where they misspelled a property.  This wouldn't be immediately compatible with some of our uses of cached properties and multiple inheritance, but that could probably be worked around.  This also would close off storing additional attributes on the objects, which is something I've sometimes done, but most users wouldn't (and likely should be discouraged anyway).  I don't think there are many performance considerations in either direction: slots occupy slightly less memory, but the number of objects in question would usually be less than a few dozen so that's pretty irrelevant; slot usage is probably slightly faster than having my protected mode on and (I think) slightly slower than having it off, but this is very unlikely to be the bottleneck.  It's a bit  annoying to keep track of which `__slots__` I need to declare, which weighs slightly in favor of my current blanket automatic solution, but there aren't that many slots, and declaring them might be a useful practice anyway.   I now mention this as another possible approach. Thanks!  Pasted here for convenience:
> An alternative approach, suggested by Matthew Feickert, would have been to use `__slots__` rather than an ordinary `__dict__` to store device attributes, which would also have the effect of raising an error if users attempt to modify unexpected attributes.  Not having a `__dict__` can make it harder to do some things like cached properties and multiple inheritance.  But in cases where such issues don't arise or can be worked around, readers facing similar challenges may find `__slots__` to be a preferable solution.
* added carriage-returns by sentence to make it easier to refer to edits by line on github.
* struck "invisibly" from "automatically invisibly".  The thought was that it makes the fact that VRML datatypes are being specified in each call invisible to the user, but you're right that "automatically" by itself already suggests pretty much that, and isn't so clunky.
* Added that it is Apache *2.0* license
* Clarified that the schism I had in mind between Python 2 and 3 communities was one *that arose* back when Python 3 opted against backwards compatibility.  I agree that it's mostly resolved now.
* Removed abbreviations like `LoC` alongside their spelled out versions like `Lines of Code`. The abbreviations are fairly commonly used so could have been helpful to a few readers, but you're right that it didn't read smoothly, and isn't really needed.
* Thanks for pointing out that the old link to the IJARS Webots paper is now broken.  That paper is also available [here](https://journals.sagepub.com/doi/pdf/10.5772/5618) if you were interested in reading it, but I'm not sure that url merits inclusion among the references, so I left it without a url for now.
* Various other little typos that your keen eye caught.  Thanks!

**Still on my to-do list:**
* add a link directly to the metric computation and compared source code files, so interested readers wouldn't need to search them out, perhaps situating stable versions on another site like zenodo
* add a workflow description to reproduce metrics.  Basically it's just pip install radon, download the two source code files and the metrics computation file, make sure the path at the top of the metrics computation file points to the directory where you situated the source files, and run it
* Sort out table-formatting and code-block-formatting issues
* Figure out what sort of "code citation" is wanted, e.g., for Webots
* Add citations for ODE, Numpy, Scipy, PIL, OpenCV
* See if I can find DOI's for citations.
* Work in a relevant image of the simulation described in section 2.

**Suggestions I'm not sure I'll adopt.**
* Saying which upcoming release of Webots this will be bundled in.  My hope is that it will be in the major 2023a release that will happen late this fall, but since I'm not in charge of Webots I can't say.
* Questioning whether the "in qualitative terms" summary adds much.  I think this summary captures the core reason why people would want to use the new API rather than the old, with much of the other content (like the immediately following bullet-list or the later code-readability metrics) offering evidence and explanation for this qualitative assessment.

stargaser commented:
@Justin-Fisher any updates on the remaining items on your to-do list for this paper?

TelosTelos committed:
citations for Numpy, OpenCv, ODE, SciPy, PIL

TelosTelos committed:
Added Radon citation, workflow for reproducing metrics

TelosTelos committed:
Added links specifically to metrics source files

TelosTelos committed:
Slightly fleshed out conclusions section.

Justin-Fisher commented:
The list is mostly done, ~~though I still need to wrestle with some table/code-block formatting issues~~.  UPDATE:  I think I got this mostly fixed by narrowing and/or manually wrapping the offending lines. 

It may also be a good idea to make a zenodo-like archival posting of the relevant source code for the metrics.  Update: DONE!

UPDATE: I added DOI's for all the references that I could find (and I fear the rest have no DOI to find).  Is there a good way to make these doi's be hyperlinked?  For now I just wrote each of them manually as a hyperlink to the relevant `dx.doi/` address.

TelosTelos committed:
Citation fixes

TelosTelos committed:
More Citation Fixes

TelosTelos committed:
Remove notes with bad formatting

TelosTelos committed:
shift codeblock formatting to fit narrow columns

TelosTelos committed:
shorten table lines to fit narrow columns

TelosTelos committed:
More informative ref-links for source code comparison versions

Justin-Fisher commented:
@scoobies build paper

scoobies commented:
building paper...

scoobies commented:
{"data"=>{"info"=>"Build for paper Justin-Fisher-master scheduled.  Note that builds are only executed if the current build attempt is more than 5 minutes old."}, "status"=>"success"}

TelosTelos committed:
Further narrow table 1

TelosTelos committed:
Even further narrow table 1

TelosTelos committed:
doi formatting

TelosTelos committed:
added all the doi's I could find

TelosTelos committed:
testing manually hyperlinking a doi

TelosTelos committed:
manually hyperlinked doi's

TelosTelos committed:
Removed More Information items that are now redundant with citations

TelosTelos committed:
Added Webots citation; more readable citation refs

TelosTelos committed:
Change links to Zenodo archival of Metrics computations and source code to compare

TelosTelos committed:
doi hyperlink

Justin-Fisher commented:
I'm not sure whether I should have expected some sort of response?

stargaser commented:
@matthewfeickert would you have time to look at this paper again?

From the editorial standpoint it looks fine.

matthewfeickert commented:
> @matthewfeickert would you have time to look at this paper again?
> 
> From the editorial standpoint it looks fine.

Yeah I've been busy with SciPy this week, but I will be able to review it by the early part of the week of 2022-07-25 (I already promised @deniederhut that I'd do this). I have two JOSS papers that I'm also late in reviewing and currently traveling for work and conferences (which take up more then the full work day).

Justin-Fisher commented:
Thanks folks.  I appreciate your time and all the helpful feedback you've already given me!

deniederhut commented:
> I have two JOSS papers that I'm also late in reviewing and currently traveling for work and conferences (which take up more then the full work day).

Busy conference season! Hope you are staying safe @matthewfeickert !

matthewfeickert commented:
> > I have two JOSS papers that I'm also late in reviewing and currently traveling for work and conferences (which take up more then the full work day).
> 
> Busy conference season! Hope you are staying safe @matthewfeickert !

Just got home on Saturday and reviewing the paper tonight. (I've been lucky to stay healthy these last 3 weeks though :+1:)

matthewfeickert commented on papers/justin_fisher/justin_fisher.rst:
```diff
@@ -0,0 +1,330 @@
+
+.. [NewAPI01] https://github.com/Justin-Fisher/new_python_api_for_webots
+
+.. [NumPy] Numerical Python (NumPy). https://www.numpy.org
```
You don't need to change this for `[NumPy]` and `[SciPy]` but if there isn't versioned code in Zenodo then I don't think you _need_ to also cite the website along with the papers. I don't mind it in any way though, so feel free to leave it. :+1:

matthewfeickert commented on papers/justin_fisher/justin_fisher.rst:
```diff
@@ -0,0 +1,330 @@
+
+.. [McC01] McCabe, TJ. "A Complexity Measure" , 2(4): 308-320. 1976.
+
+.. [Mic01] Michel, O. "Webots: Professional Mobile Robot Simulation. *Journal of Advanced Robotics Systems.* 1(1): 39-42. 2004. `doi: 10.5772/5618 <dx.doi.org/10.5772/5618>`_.
```
Same issue with a broken DOI link so maybe try https://doi.org/10.5772/5618 ?
```suggestion
.. [Mic01] Michel, O. "Webots: Professional Mobile Robot Simulation. *Journal of Advanced Robotics Systems.* 1(1): 39-42. 2004. `doi: 10.5772/5618 <https://doi.org/10.5772/5618>`_.
```

Justin-Fisher replied:
That's odd that writing the doi links like this seems to work fine in some cases (e.g. the Buse and Hindle papers), but not in others, like this one, and I can't tell what's causing the difference.  (As a first guess, it seems to work when the url contains a fully alpha word just after the first slash, and not when it contains something purely numeric there, though I don't know why that would matter.)  I'll try adding the explicit `https://`.

Justin-Fisher replied:
That does seem to work, so I've prepended `https://` to all of them, even the ones that already worked.  Thanks for catching this and suggesting a workable fix!

matthewfeickert commented on papers/justin_fisher/justin_fisher.rst:
```diff
@@ -0,0 +1,330 @@
+
+.. [ODE] Open Dynamics Engine. https://www.ode.org/
+
+.. [Oman01] Oman, P and J Hagemeister. "Metrics for assessing a software system's maintainability," *Proceedings Conference on Software Maintenance*, 337-44. 1992.
```
There's a DOI for this paper https://doi.org/10.1109/ICSM.1992.242525

```suggestion
.. [Oman01] Oman, P and J Hagemeister. "Metrics for assessing a software system's maintainability," *Proceedings Conference on Software Maintenance*, 337-44. 1992. `doi: 10.1109/ICSM.1992.242525 <https://doi.org/10.1109/ICSM.1992.242525>`_.
```

Justin-Fisher replied:
Thanks for finding this.  Now added!

matthewfeickert commented on papers/justin_fisher/justin_fisher.rst:
```diff
@@ -0,0 +1,330 @@
+
+.. [Radon] Radon. https://radon.readthedocs.io/en/latest/index.html
+
+.. [Sca01] Scalabrino, S, M Linares-Vasquez, R Oliveto and D Poshyvanyk. "A Comprehensive Model for Code Readability." *Jounal of Software: Evolution and Process*, 1-29. 2017. `doi: 10.1002/smr <dx.doi.org/10.1002/smr>`_.
```
The paper itself https://www.cs.wm.edu/~denys/pubs/JSEP'18-ReadabilityPaper.pdf gives the wrong DOI (missing the actual identifying numbers), so not your fault, but the DOI is actually https://doi.org/10.1002/smr.1958

```suggestion
.. [Sca01] Scalabrino, S, M Linares-Vasquez, R Oliveto and D Poshyvanyk. "A Comprehensive Model for Code Readability." *Jounal of Software: Evolution and Process*, 1-29. 2017. `doi: 10.1002/smr.1958 <https://doi.org/10.1002/smr.1958>`_.
```

Justin-Fisher replied:
Another good catch and solution. Thanks!

matthewfeickert commented on papers/justin_fisher/justin_fisher.rst:
```diff
@@ -0,0 +1,330 @@
+
+.. [SDTOld] https://github.com/Justin-Fisher/new_python_api_for_webots/blob/d180bcc7f505f8168246bee379f8067dfaf373ea/webots_new_python_api_samples/controllers/supervisor_draw_trail_python/supervisor_draw_trail_old_api_bare_bones.py
+
+.. [Vir01] Virtanen, P, R. Gommers, T. Oliphant, et al. SciPy 1.0: Fundamental Algorithms for Scientific Computing in Python. *Nature Methods*, 17(3), 261-72. 2020.  `doi: 10.1038/s41592-019-0686-2 <dx.doi.org/10.1038/s41592-019-0686-2>`_.
```
Same issue wrt DOIs in the render for https://doi.org/10.1038/s41592-019-0686-2

```suggestion
.. [Vir01] Virtanen, P, R. Gommers, T. Oliphant, et al. SciPy 1.0: Fundamental Algorithms for Scientific Computing in Python. *Nature Methods*, 17(3), 261-72. 2020. `doi: 10.1038/s41592-019-0686-2 <https://doi.org/10.1038/s41592-019-0686-2>`_.
```

matthewfeickert reviewed (changes requested):
Thanks for your patience @Justin-Fisher @deniederhut and for a very nice job of addressing things and improving the document quite a bit. My review comments as code review are super minor and most of them are just corrections to URLs so that DOIs should have valid URLs in the render. :+1: 

Also the top level `justin_fisher.rst` that is for another paper still needs to be removed.

More formally, my updated review is the following:

## Written Quality

- [x] Prose should be written in English.
- [x] Prose should clearly and consistently communicate the narrative (e.g. consistent tense and person; use of active voice; etc)
- [x] The written style should convey information that furthers the knowledge or research of the reader.
- [x] Due to the interdisciplinary nature of SciPy, highly domain-specific jargon should be avoided or explained where possible.

## Technical Content

- [x] The technical content should be scientifically sound.
- [x] Computational content should, likewise, be verified.
- [x] The work should describe the development or use of python software for approaching a problem in a domain within the scope of the conference.
   * Yes, it seems to be a nice use case combining robotics and education.

## Novelty and Impact

- [x] Assume that the novelty of the conceptual work described in the paper has been assessed by the Program Committee.
  The paper should advance the state of a scientific domain, the practice of scientific computing itself, or another subject area within the scope of the conference.
   * The paper provides work on a Pythonic API to a robotics simulator that provides an alternative to the C-like Python API/bindings that are currently provided. This work is substantial and combined with the author's use of it in University level educational settings it is well within the scope of the conference.

## Verifiability

- [x] Software descriptions should be accompanied by references to or examples of representative source code.
- [x] Source code essential to the conclusions of the paper should be made available to the reader.
   * As I have yet to fully replicate the results I can't say with certainty that the source code is sufficient, but for the time being I am willing to say that as it exists publicly that it is hypothetically replicable.
- [x] Data sources should be identified (e.g., with citation to a persistent DOI).
   * In this situation data include source code and dependencies. The provided source code should ideally be cited at a specific tag, and not just linked, and ideally be given a DOI (i.e. [with Zenodo](https://zenodo.org/)), but having commits is probably enough for this paper.
- [ ] Analysis should be accompanied by a workflow description sufficient to reproduce the results.
  * Formal Analysis workflow instructions are not provided, but I won't block moving forward on this, so this can be ignored for the most part.
  
## Other Requirements

- [x] All mentioned software should be formally and consistently cited wherever possible.
   * There are some DOI corrections to be made, but things are basically there.
- [x] Acronyms should be spelled out upon first mention.
- [x] License conditions on images and figures must be respected (Creative Commons, etc.).
- [x] Mathematical and other symbols should be defined.
- [x] Definitions should include consistent units where appropriate.
- [x] Avoid long paragraphs of your rich text file to enable reviewers to target specific sections with their comments.
  GitHub comments are per line of text.
- [x] Avoid custom LaTeX markup where possible.

### Length

- [x] The compiled version should be no longer than 8 pages, including figures, source code, appendices. References are excluded from this limit.

### Figures

- [ ] All figures and tables should have captions.
   * None of the tables have sufficient captions. Captions should provide sufficient information that a reader is able to understand the core components and relevance of the figure or table. I'll defer to the editor (@stargaser) here, but it would be nice to see these improved a bit.
- [x] Figure text should be of a readable size.
- [x] Images and figures should be reasonably sized and formatted for viewing online; typically a few hundred kilobytes and less than 1 MB.

### Code Snippets

- [x] Code snippets should be formatted to fit inside a single column without overflow. Please make sure to check the build server's copy of your paper; it's the authoritative view of how your paper will appear.
- [x] If not in Python, the language in the code snippet should be mentioned.
- [x] Code snippets should follow a common style guide. PEP8 is preferred.
   * All the code snippets are incomplete segments and not actually runnable, so while it is unclear if a formatting style is being used it isn't very relevant for 1-liners, so this can be skipped.

matthewfeickert commented on papers/justin_fisher/justin_fisher.rst:
```diff
@@ -0,0 +1,330 @@
+
+.. [Hal01] Halstead, M. *Elements of software science.* Elsevier New York. 1977.
+
+.. [Har01] Harris, C., K. Millman, S. van der Walt, et al. Array programming with NumPy. *Nature* 585, 357–62. 2020. `doi: 10.1038/s41586-020-2649-2 <dx.doi.org/10.1038/s41586-020-2649-2>`_.
```
For some reason the DOI isn't linking properly in the render, so maybe try https://doi.org/10.1038/s41586-020-2649-2 ? (This was supposed to be posted first but somehow got dropped in the review comments, sorry for confusion.)

```suggestion
.. [Har01] Harris, C., K. Millman, S. van der Walt, et al. Array programming with NumPy. *Nature* 585, 357–62. 2020. `doi: 10.1038/s41586-020-2649-2 <https://doi.org/10.1038/s41586-020-2649-2>`_.
```

deniederhut commented:
Thanks @matthewfeickert ! @Justin-Fisher friendly warning that we'll be publishing the final proceedings this weekend, so we'll need to have changes to your paper finalized before then.

matthewfeickert commented:
> @Justin-Fisher friendly warning that we'll be publishing the final proceedings this weekend, so we'll need to have changes to your paper finalized before then.

@Justin-Fisher as my last review was delayed and so hasn't left you with a huge amount of time hopefully the suggested DOI fixes and additions will be easy to implement and check to make sure those render out okay. Let me know if you have any questions, but as I'm currently traveling again you might want to tag @deniederhut as well.

TelosTelos committed:
add https to broken doi hyperlink

TelosTelos committed:
add https:// to all doi hyperlinks

TelosTelos committed:
added doi for Oman paper

TelosTelos committed:
fix Scalabrino doi

TelosTelos committed:
attempting caption for table 1

TelosTelos committed:
attempting to left align and boldface caption

TelosTelos committed:
captions for other tables

TelosTelos committed:
fix labels for tables

TelosTelos committed:
paring back alignment change to try to fix failure

TelosTelos committed:
Further narrow Table 3 now that captions have column-wrapped it to the side where it must be narrower

TelosTelos committed:
Yet another attempt to narrow

TelosTelos committed:
Fix table 2 linewrap

TelosTelos committed:
remove remnant file from old conference

Justin-Fisher commented:
Ok, I think I've finished making all the changes that @matthewfeickert suggested.  Thanks for all the good catches and for suggesting workable fixes!

I'm not totally happy with how the table captions look:  (1) it insists on adding a separate redundant line saying, e.g., `TABLE 1`, and (2) the captions would look better if they were left-justified rather than centered.  (I've tried using `:align: left` in the table but that doesn't seem to affect the caption.)  But I don't know of a way to control either of these?  Neither of these is a big deal, and it's probably fine as is.

(Since Matthew is traveling I'm following his suggestion to tag @deniederhut as well.)

deniederhut commented:
@Justin-Fisher  latex is handling the table references for you. When you type 
```
:label:`metrictable`
```
it replaces it with the appropriate table number, based on the order of the tables in the paper. Probably the easiest thing is to remove `Table 1` from 

```
 .. table:: **Table 1. Length and Complexity Metrics.** 
```

deniederhut commented:
@stargaser do you have time to take a final pass over the paper before we publish?

TelosTelos committed:
remove redundant "Table 1"

Justin-Fisher commented:
Thanks, that removes the redundancy, and is close enough to the "ideal" formatting for me.

stargaser commented:
I've made a final pass over the paper. It looks acceptable to me.

matthewfeickert reviewed (approved):
LGTM @Justin-Fisher! :+1: Thanks for your contributions to SciPy!

@deniederhut @stargaser I don't want to delay things at all and I don't want to ask @Justin-Fisher to do any more work here, but as Justin pointed out for the table caption formatting

> (2) the captions would look better if they were left-justified rather than centered. (I've tried using :align: left in the table but that doesn't seem to affect the caption.) 

and I agree with him here. I'm not sure if either of you know how to fix this and if you can push changes for it. If not, then not a big deal.

deniederhut commented:
Might have to be something we tackle next year. Thanks everyone!

Justin-Fisher commented:
Thanks everyone, especially @matthewfeickert for your many good catches and helpful suggestions.  If only all reviewers were like you!
